### ISSUE TITLE
Forge's PR Section is quite slow

### ISSUE THREAD

**Issue by rgrinberg** [magit/forge#103]
Here's a profile report in a very happy optimistic case:

```
- command-execute                                                 320  94%
 - call-interactively                                             320  94%
  - funcall-interactively                                         318  94%
   - magit-status                                                 316  93%
    - magit-status-internal                                       316  93%
     - magit-mode-setup                                           316  93%
      - magit-mode-setup-internal                                 316  93%
       - magit-refresh-buffer                                     304  89%
        - apply                                                   303  89%
         - magit-status-refresh-buffer                            303  89%
          - magit-run-section-hook                                302  89%
           - apply                                                302  89%
            - forge-insert-pullreqs                               272  80%
             - #<compiled 0x446e2551>                             270  79%
              - forge-insert-pullreq                              270  79%
               - forge--insert-pullreq-commits                    202  59%
                - #<compiled 0x4804d225>                          159  47%
                 - magit-insert-log                               159  47%
                  - magit-git-wash                                159  47%
                   - #<compiled 0x457a6c11>                       112  33%
                    - apply                                       112  33%
                     - forge-insert-issues                        101  29%
                      - forge-list-recent-topics                  101  29%
                       - apply                                    101  29%
                        - #<compiled 0x478e9f79>                   89  26%
                         - closql--remake-instance                 89  26%
                          - apply                                  89  26%
                           - #<compiled 0x478ed4c5>                 89  26%
                            - closql--resolve-slots                 89  26%
                             + apply                               89  26%
                        + #<compiled 0x47922039>                   10   2%
                        + #<compiled 0x4790be15>                    2   0%
                     + magit-log-wash-log                          11   3%
                   + magit-git-insert                              47  13%
                - forge--pullreq-ref                               43  12%
                 + magit-rev-verify                                33   9%
                 + forge--pullreq-ref-1                            10   2%
               - forge--format-topic-labels                        68  20%
                - closql--iref                                     68  20%
                 - closql--table-columns                           35  10%
                  - emacsql                                        35  10%
                   - apply                                         35  10%
                    + #<compiled 0x447a39b9>                       35  10%
                 - emacsql                                         32   9%
                  + apply                                          32   9%
             + forge-get-repository                                 1   0%
             + forge-list-recent-topics                             1   0%
            + magit-insert-status-headers                           7   2%
            + magit-insert-worktrees                                5   1%
            + magit-insert-unpushed-to-upstream-or-recent                  5   1%
            + magit-insert-unpushed-to-pushremote                   3   0%
            + magit-insert-unstaged-changes                         2   0%
            + magit-insert-stashes                                  2   0%
            + magit-insert-merge-log                                1   0%
            + magit-insert-rebase-sequence                          1   0%
            + magit-insert-untracked-files                          1   0%
            + magit-insert-staged-changes                           1   0%
            + magit-insert-unpulled-from-pushremote                  1   0%
            + magit-insert-unpulled-from-upstream                   1   0%
          + magit-git-exit-code                                     1   0%
       + magit-status-mode                                          8   2%
       + run-hooks                                                  1   0%
   + profiler-report                                                2   0%
  - byte-code                                                       2   0%
   + magit-toplevel                                                 2   0%
+ timer-event-handler                                              12   3%
+ ...                                                               5   1%
+ redisplay_internal (C function)                                   1   0%
```

In practice, I have to wait for at least a few seconds for some my larger repos to render. Note that I would be quite fine with the commits for every PR not showing up if this feature was faster.

Lastly: Thank you very much for making forge. I could never get magithub to work the way I wanted because of its performance issues. I have high hopes for forge solving all these problems.

**Comment by vermiculus** [comment]
I'm surprised this is being done all at once. It was my understanding that magit-section now waits to run the body code until it's actually expanded.

If I'm interpreting the profile information correctly, it looks like most time is spent re-creating objects from the database.

As for magithub performance, it was a mistake to try and provide the topic body in the status buffer. You can fully customize what information is shown with the sections hook. But, this issue is not about magithub and alas I'll be moving most functionality into forge once it's stabilized more.

**Comment by tarsius** [comment]
Did you give `forge-topic-list-limit` a try?

**Comment by rgrinberg** [comment]
I read the docs for this variable, but I'm not sure how should I set it. Suppose I want to show a maximum of 10 issues and 10 pr's, without any of the commits for the PR's visible. How should I set this variable?

**Comment by vermiculus** [comment]
You can't skip the commit-log with this variable, but I'm still unsure of why magit-section is behaving this way (unless I misunderstand how magit-section is supposed to behave in the first place).

You can't set a maximum, either – from the docstring, all unread topics will always be shown.

**Comment by tarsius** [comment]
```
               - forge--insert-pullreq-commits                    202  59%
                - #<compiled 0x4804d225>                          159  47%
                 - magit-insert-log                               159  47%
```

The former is supposed to be called on every fresh, the latter is not. I inserted a debug statement inside the `magit-insert-section-body` inside `forge--insert-pullreq-commits` and that confirmed that the body of the macro is not executed until the section is expanded, as intended.

Are all the pr sections expanded for you? Try inserting such a debug statement too, when does it get executed?

**Comment by rgrinberg** [comment]
Okay, folding the Issues/Pr's sections indeed makes `magit-status` again. But I fear that this really cuts down on the usefulness of the feature for me. I basically have to remember to hide the section before closing the buffer.

To give another profile of how slow it on OSX:

```
- ...                                                            2396  98%
 - #<compiled 0x45b54155>                                        2191  90%
  - closql-get                                                   2191  90%
   - apply                                                       2191  90%
    - #<compiled 0x4749475d>                                     2191  90%
     - closql--remake-instance                                   2158  88%
      - apply                                                    2158  88%
       - #<compiled 0x474984a5>                                  2158  88%
        - closql--resolve-slots                                  2158  88%
         - apply                                                 2158  88%
          - #<compiled 0x474984d1>                               2158  88%
           - slot-boundp                                         2158  88%
            - eieio-oref                                         2158  88%
             - apply                                             2158  88%
              - eieio-oref--closql-oref                          2117  87%
               - closql-oref                                     2117  87%
                - emacsql                                        1940  79%
                 - apply                                         1940  79%
                  - #<compiled 0x40933a23>                       1940  79%
                   - apply                                       1940  79%
                    - #<compiled 0x47494651>                     1940  79%
                     - apply                                     1928  79%
                      - #<compiled 0x481b5671>                   1928  79%
                       - apply                                   1928  79%
                        - #<compiled 0x4748dbed>                 1928  79%
                         - emacsql-wait                          1717  70%
                          - apply                                1717  70%
                           - #<compiled 0x4748abd1>               1717  70%
                            + accept-process-output                  5   0%
                         + emacsql-parse                          139   5%
                         + emacsql-send-message                    70   2%
                         + apply                                    2   0%
                     + mapcar                                      12   0%
                + mapcar                                          105   4%
                + closql--table-columns                            72   2%
              + #<compiled 0x45b56b55>                             41   1%
     + emacsql                                                     33   1%
   Automatic GC                                                   156   6%
 + #<compiled 0x44407bf1>                                          46   1%
 + run-hooks                                                        1   0%
 + profiler-calltree-sort                                           1   0%
 + page-break-lines--update-display-tables                          1   0%
+ command-execute                                                  21   0%
+ timer-event-handler                                               9   0%
```

That's just to load 34 PR's!

**Comment by tarsius** [comment]
> I basically have to remember to hide the section before closing the buffer.

Why are they *all* expanded in the first place? They are supposed to be collapsed by default. Isn't that the case for you? But I guess you explicitly expand them all. Doesn't that make the information rather anyway? Personally I only expand the one pr I am interested in and assumed others would do the same.

**Comment by rgrinberg** [comment]
> Why are they _all_ expanded in the first place? They are supposed to be collapsed by default. Isn't that the case for you? But I guess you explicitly expand them all

I do not expand them all. I only expand the PR's section. Even that is taking 3-4 seconds on my machine (2017 MBP, not at all an old machine).

**Comment by tarsius** [comment]
Just to be sure: When the status buffer contains:

```
Pull requests (8)
#47 Display inline images in topics
90b5473 vermiculus/inline-images 47 fixup! Display inline images in topics
75a8397 Display inline images in topics

#90 Generalize browse, visit commands and introduce copy-URL command
#100 Use EMake for the build
#97 Generalize topic list insertion
#95 Provide context when creating posts
#53 Bind RET to follow Markdown links in issue buffers
#74 Add "Forge" to magit-dispatch-popup
#77 Bind forge-list-visit-* in forge-*-list-mode-map w/ magit convention
```

instead of just

```
Pull requests (8)
#47 Display inline images in topics
#90 Generalize browse, visit commands and introduce copy-URL command
#100 Use EMake for the build
#97 Generalize topic list insertion
#95 Provide context when creating posts
#53 Bind RET to follow Markdown links in issue buffers
#74 Add "Forge" to magit-dispatch-popup
#77 Bind forge-list-visit-* in forge-*-list-mode-map w/ magit convention
```

then that makes refreshing several seconds slower?

**Comment by rgrinberg** [comment]
> then that makes refreshing several seconds slower?

It's neither.

I'm sorry for wasting everyone's time. I don't mean to make it so hard to understand what the problem is.

Let me try again.

1) Situation 1: The PR's section is open and I see only a list of PR's.

In this situation, opening the status screen is absurdly slow. Around 5 seconds for small repos.

2) Situation 2: The PR's section is closed and I don't see any PR's.

In this situation, opening the status screen is "fast". I wish it was faster still, but it's tolerable.

I made a little screencast to show the 2 modes:

https://photos.google.com/share/AF1QipNgaAKnqdJZ98y8eBHiEDvS4AMOAX11_k7E8a7mSNxmC_NjUEu5nNIUjPT83c409Q?key=cnFyQnl6azM5NFQ5NkV1MU9NVGdVRElHbnB1U2tB

The quality is horrible, but it should be enough to describe the problem. In the beginning, things are comically slow. When I hide the PR's section altogether, performance becomes reasonable.

**Comment by Linuus** [comment]
I just want to chime in saying I see the exact same issues.

When the status buffer looks like this:

```
> Pull requests (8)
```

`magit-status` runs OK. Not fast, but fine :)

When it looks like this:

```
Pull requests (8)
#47 Display inline images in topics
#90 Generalize browse, visit commands and introduce copy-URL command
#100 Use EMake for the build
#97 Generalize topic list insertion
#95 Provide context when creating posts
#53 Bind RET to follow Markdown links in issue buffers
#74 Add "Forge" to magit-dispatch-popup
#77 Bind forge-list-visit-* in forge-*-list-mode-map w/ magit convention
```
It takes maybe 10-15s to run `magit-status`.

I have maybe 15 PR's in my Pull Request list.

**Comment by tarsius** [comment]
I was experiencing this too today and the remembered that I have seen it before.

But I couldn't figure out what was going on. After recompiling the issue was gone in a new Emacs instance. However it was also still gone after removing all byte-code files.

**Comment by tarsius** [comment]
I briefly though that #97 was to blame, but it doesn't seem to cause the the multi-second delay.

However it does seem to cause a much smaller but still noticeable delay, which I cannot explain yet.

**Comment by rgrinberg** [comment]
> And none of those 34 pull requests are expanded, yeah? How long does it take you to run the following?

It doesn't seem very slow to me:

```
(benchmark-run (forge-list-recent-topics (forge-get-repository t) 'pullreq))
(0.037178 0 0.0)
```

**Comment by rgrinberg** [comment]
Another data point:

Expanding out the list of issues doesn't seem nearly as slow. I still notice a delay, but it's nowhere near as bad.

**Comment by rgrinberg** [comment]
I've updated to the latest version of forge and things seem to be much faster. It's still not as fast as I'd like (1-2 seconds delay), but it's manageable.

I'm putting up one last profile:
```
- command-execute                                                 194  58%
 - call-interactively                                             194  58%
  - funcall-interactively                                         193  58%
   - magit-status                                                 192  58%
    - magit-status-internal                                       192  58%
     - magit-mode-setup                                           192  58%
      - magit-mode-setup-internal                                 192  58%
       - magit-refresh-buffer                                     183  55%
        - apply                                                   182  55%
         - magit-status-refresh-buffer                            182  55%
          - magit-run-section-hook                                181  55%
           - apply                                                181  55%
            - forge-insert-pullreqs                               104  31%
             - forge-insert-topics                                 90  27%
              - #<compiled 0x44580751>                             89  27%
               - forge-insert-topic                                89  27%
                - forge--insert-topic-contents                     88  26%
                 - apply                                           88  26%
                  - #<compiled 0x41068d5d>                         88  26%
                   - apply                                         88  26%
                    - #<compiled 0x47826c75>                       59  17%
                     - forge--format-topic-labels                  59  17%
                      - closql--iref                               59  17%
                       - closql--table-columns                     36  10%
                        - emacsql                                  36  10%
                         - apply                                   36  10%
                          - #<compiled 0x475e8525>                 36  10%
                           - apply                                 36  10%
                            - #<compiled 0x46d9c7b5>                 36  10%
                             - apply                               36  10%
                              - #<compiled 0x4459b709>                 36  10%
                               - apply                             36  10%
                                - #<compiled 0x46d97349>                 36  10%
                                 + emacsql-wait                    20   6%
                                 + emacsql-send-message                 16   4%
                       + emacsql                                   23   6%
                    + #<compiled 0x46def4dd>                       29   8%
                + run-hook-with-args-until-success                  1   0%
              + magit-section                                       1   0%
             + forge-list-recent-topics                            12   3%
             + forge-get-repository                                 1   0%
            - magit-insert-unpushed-to-upstream-or-recent                 45  13%
             - magit-insert-unpushed-to-upstream                   44  13%
              - magit-insert-log                                   42  12%
               - magit-git-wash                                    42  12%
                - #<compiled 0x44a7b579>                           41  12%
                 - apply                                           41  12%
                  - magit-log-wash-log                             41  12%
                   - magit-wash-sequence                           41  12%
                    - #<compiled 0x44a7b655>                       41  12%
                     - apply                                       41  12%
                      - magit-log-wash-rev                         34  10%
                       + magit-format-ref-labels                    6   1%
                       + magit-log-format-margin                    6   1%
                       + run-hook-with-args-until-success                  4   1%
                       + put-text-property                          4   1%
                         jit-lock-after-change                      2   0%
                       + magit-delete-line                          2   0%
                         magit-insert-child-count                   1   0%
                + magit-git-insert                                  1   0%
              + magit-git-success                                   1   0%
              + put-text-property                                   1   0%
             + magit-rev-parse                                      1   0%
            + magit-insert-unpulled-from-upstream                  12   3%
            + magit-insert-status-headers                           7   2%
            + magit-insert-worktrees                                3   0%
            + magit-insert-unpushed-to-pushremote                   3   0%
            + magit-insert-merge-log                                1   0%
            + magit-insert-untracked-files                          1   0%
            + magit-insert-unstaged-changes                         1   0%
            + magit-insert-staged-changes                           1   0%
            + magit-insert-stashes                                  1   0%
            + magit-insert-unpulled-from-pushremote                  1   0%
          + magit-git-exit-code                                     1   0%
       + magit-status-mode                                          7   2%
       + magit-display-buffer                                       1   0%
       + run-hooks                                                  1   0%
   + profiler-report                                                1   0%
  + byte-code                                                       1   0%
- ...                                                             133  40%
   Automatic GC                                                   133  40%
```

I'll close this issue for now. Thanks all
